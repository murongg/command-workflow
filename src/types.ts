export interface StepTags {
  [x: string]: (() => string) | string
}
export interface Step {
  command: string
  error?: (error: Error) => void
  tags?: StepTags
  before?: (command: string, tags: Record<string, any>) => string | undefined
  after?: (command: string, buffer: Buffer) => void
}

