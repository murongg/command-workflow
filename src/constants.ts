export const DEFAULT_CONFIG_SUFFIXES = [
  '.config.js',
  '.config.mjs',
  '.config.ts',
  '.config.cjs',
  '.config.mts',
  '.config.cts',
]

export const DEFAULT_CONFIG_PREFIXES = [
  'cwf',
  'command-workflow',
  'command',
  'workflow',
]

export const getDefaultConfigPrefixes = () => {
  const files = []
  for (const prefix of DEFAULT_CONFIG_PREFIXES) {
    for (const suffix of DEFAULT_CONFIG_SUFFIXES)
      files.push(`${prefix}${suffix}`)
  }
  return files
}
