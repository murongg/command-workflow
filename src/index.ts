export { defineConfig } from './config'
export { start } from './run'
export type { StepTags, Step } from './types'
export type { UserConfig, UserConfigFn, UserConfigMap, UserConfigExport } from './config'
