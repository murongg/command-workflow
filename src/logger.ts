/* eslint no-console: 0 */
// fork vite logger.ts
import colors from 'picocolors'

export type LogType = 'error' | 'warn' | 'info'
export type LogLevel = LogType | 'silent'
export interface Logger {
  info(msg: string, options?: LogOptions): void
  warn(msg: string, options?: LogOptions): void
  warnOnce(msg: string, options?: LogOptions): void
  error(msg: string, options?: LogErrorOptions): void
  hasWarned: boolean
}

export interface LogOptions {
  clear?: boolean
  timestamp?: boolean
}

export interface LogErrorOptions extends LogOptions {
  error?: Error | null
}

export const LogLevels: Record<LogLevel, number> = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
}

export interface LoggerOptions {
  prefix?: string
}

export function createLogger(
  level: LogLevel = 'info',
  options: LoggerOptions = {},
): Logger {
  const { prefix = '[CWF]' } = options
  const thresh = LogLevels[level]

  function output(type: LogType, msg: string, options: LogErrorOptions = {}) {
    if (thresh >= LogLevels[type]) {
      const method = type === 'info' ? 'log' : type
      const format = () => {
        if (options.timestamp) {
          const tag
            = type === 'info'
              ? colors.cyan(colors.bold(prefix))
              : type === 'warn'
                ? colors.yellow(colors.bold(prefix))
                : colors.red(colors.bold(prefix))
          return `${colors.dim(new Date().toLocaleTimeString())} ${tag} ${msg}`
        }
        else {
          return msg
        }
      }
      if (options.error)
        msg += `\n${options.error.stack}`

      console[method](format())
    }
  }

  const warnedMessages = new Set<string>()

  const logger: Logger = {
    hasWarned: false,
    info(msg, opts) {
      output('info', msg, opts)
    },
    warn(msg, opts) {
      logger.hasWarned = true
      output('warn', msg, opts)
    },
    warnOnce(msg, opts) {
      if (warnedMessages.has(msg))
        return
      logger.hasWarned = true
      output('warn', msg, opts)
      warnedMessages.add(msg)
    },
    error(msg, opts) {
      logger.hasWarned = true
      output('error', msg, opts)
    },
  }

  return logger
}
