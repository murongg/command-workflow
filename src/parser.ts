import type { StepTags } from './types'

/**
 * e.g. git tag v#{tag} -m v#{message}
 * parse to git tag v1.0.0 -m v1.0.0
 * @param template
 * @param tags
 * @returns
 */
export function parserTemplateTag(template: string, tags: StepTags): string {
  return template.replace(/#\{(\w+)\}/g, (_, key) => {
    const value = tags[key]
    if (typeof value === 'function')
      return value()

    return value
  })
}
