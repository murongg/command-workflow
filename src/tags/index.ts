import { getTimestamp } from './date'
import { getCurrentGitBranch, getCurrentGitCommit, getCurrentGitCommitMessage, getCurrentGitRepo, getCurrentGitRepoName, getCurrentGitRepoOwner, getCurrentGitRepoUrl, getCurrentGitTag, getGitUserEmail, getGitUserName } from './git'

export const TAGS_MAP = {
  timestamp: getTimestamp,
  current_git_branch: getCurrentGitBranch,
  current_git_commit: getCurrentGitCommit,
  current_git_commit_message: getCurrentGitCommitMessage,
  current_git_tag: getCurrentGitTag,
  current_git_repo: getCurrentGitRepo,
  current_git_repo_url: getCurrentGitRepoUrl,
  current_git_repo_name: getCurrentGitRepoName,
  current_git_repo_owner: getCurrentGitRepoOwner,
  git_user_name: getGitUserName,
  git_user_email: getGitUserEmail,
}
