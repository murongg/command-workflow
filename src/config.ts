import { join, resolve } from 'node:path'
import { existsSync } from 'node:fs'
import { importModule } from 'local-pkg'
import { getDefaultConfigPrefixes } from './constants'
import type { Step } from './types'
import type { LogLevel } from './logger'
import { createLogger } from './logger'
import { isWindows, randomUniqueKey } from './utils'

export interface UserConfig {
  /**
   * Log level.
   * @default 'info'
   */
  logLevel?: LogLevel
  /**
   * Whether to skip error.
   * @default false
   */
  isSkipError?: boolean
  /**
   * When throw error whether to break.
   * @default false
   */
  isThrowErrorBreak?: boolean
  /**
   * Steps.
   */
  steps: (Step | string)[]
}

export type UserConfigFn = (...args: any[]) => UserConfig | Promise<UserConfig>

export interface UserConfigMap {
  default: UserConfig | UserConfigFn | null
  [x: string]: UserConfig | UserConfigFn | null
}
export type UserConfigExport = UserConfig | Promise<UserConfig> | UserConfigMap | Promise<UserConfigMap> | UserConfigFn

export function defineConfig(config: UserConfig): UserConfig
export function defineConfig(config: Promise<UserConfig>): Promise<UserConfig>
export function defineConfig(config: UserConfigMap): UserConfigMap
export function defineConfig(config: Promise<UserConfigMap>): Promise<UserConfigMap>
export function defineConfig(config: UserConfigFn): UserConfigFn
export function defineConfig(config: UserConfigExport): UserConfigExport {
  return config
}

type UserConfigOnlyStep = Omit<UserConfig, 'steps'> & { steps: Step[] }
interface UserConfigMapOmitFn {
  default: UserConfigOnlyStep | null
  [x: string]: UserConfigOnlyStep | null
}

const config: UserConfigMapOmitFn = {
  default: null,
}
export async function getConfig(key?: string, configFile?: string, configRoot?: string) {
  if (config.default && !key)
    return config
  const configTmp = await loadConfigFromFile(configFile, configRoot)
  let realConfig: UserConfig = {
    steps: [],
  }
  if (Reflect.has(configTmp || {}, 'steps')) {
    realConfig = typeof configTmp === 'function' ? await configTmp() : configTmp as UserConfig
  }
  else if (Reflect.has(configTmp || {}, key || 'default')) {
    const value = Reflect.get(configTmp || {}, key || 'default')
    realConfig = typeof value === 'function' ? await value() : value
  }
  else {
    realConfig = configTmp as UserConfig
  }
  config.default = {
    steps: [],
    logLevel: realConfig?.logLevel || 'info',
    isSkipError: realConfig?.isSkipError || false,
    isThrowErrorBreak: realConfig?.isThrowErrorBreak || false,
  }
  config.default.steps = realConfig?.steps.map((item) => {
    if (typeof item === 'string') {
      return {
        command: item,
        unikey: randomUniqueKey(),
      }
    }
    else if (typeof item === 'object') {
      return {
        ...item,
        unikey: item.unikey || randomUniqueKey(),
      }
    }
    return item
  })

  if (!config.default?.steps)
    createLogger(config.default?.logLevel).error('No steps found in config file.', { timestamp: true })
  return config
}

export async function loadConfigFromFile(configFile?: string, configRoot: string = process.cwd()): Promise<UserConfig | UserConfigFn | UserConfigMap | null> {
  let resolvedPath = ''
  if (configFile) {
    resolvedPath = resolve(join(configRoot, configFile))
  }
  else {
    const configFiles = getDefaultConfigPrefixes()

    for (const filename of configFiles) {
      const filePath = resolve(configRoot, filename)
      if (!existsSync(filePath))
        continue

      resolvedPath = filePath
      break
    }
  }
  if (!resolvedPath)
    return null

  if (isWindows)
    resolvedPath = `file:///${resolvedPath.replace(/\\/g, '/')}`

  try {
    const config = await importModule(resolvedPath)
    return typeof config.default === 'function' ? config.default() : config.default
  }
  catch (error) {
    createLogger('error').error('Failed to load config file.', { error: error as Error, timestamp: true })
    return null
  }
}
