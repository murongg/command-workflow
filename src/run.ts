import { execSync } from 'child_process'
import { loadConfigFromFile } from './config'
import { parserTemplateTag } from './parser'

export async function run() {
  const config = await loadConfigFromFile()
  if (!config?.steps)
    throw new Error('No steps found in config file')
  if (config && config.steps) {
    const commands: string[] = []
    for (const step of config.steps) {
      const cmd = parserTemplateTag(step.command, step.tags || {})
      commands.push(cmd)
      try {
        execSync(cmd, { stdio: 'inherit' })
      }
      catch (error) {
        console.error(error)
      }
    }
  }
}
