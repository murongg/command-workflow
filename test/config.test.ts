import { resolve } from 'path'
import { describe, expect, it } from 'vitest'
import { type UserConfig, defineConfig, getConfig, loadConfigFromFile } from '../src/config'

describe('describe', () => {
  it('test defineConfig', () => {
    const config: UserConfig = {
      logLevel: 'info',
      steps: [{
        command: 'git tag v#{tag} -m #{message}',
        tags: {
          tag: 'tag',
          message: 'message',
        },
      }],
    }
    expect(defineConfig(config)).toMatchSnapshot()
  })

  it('test loadConfigFromFile', async () => {
    // test default
    const cwd = resolve(__dirname)
    const output = await loadConfigFromFile(undefined, cwd)
    const res = {
      logLevel: 'info',
      steps: [{
        command: 'git tag v#{tag} -m #{message}',
        tags: {
          tag: 'tag',
          message: 'message',
        },
      }],
    }
    expect(output).toEqual(res)
    // test file
    const output2 = await loadConfigFromFile('cwf.config.js', __dirname)
    expect(output2).toEqual(res)
    // test error
    const output3 = await loadConfigFromFile('command.config.js', __dirname)
    expect(output3).toThrowError(Error)
    // test null
    const output4 = await loadConfigFromFile()
    expect(output4).toBe(null)
  })

  it('test getConfig', async () => {
    const cwd = resolve(__dirname)
    const output = await getConfig(undefined, undefined, cwd)
    const res = {
      logLevel: 'info',
      isSkipError: false,
      isThrowErrorBreak: false,
      steps: [{
        command: 'git tag v#{tag} -m #{message}',
        tags: {
          tag: 'tag',
          message: 'message',
        },
      }],
    }
    // remove unikey
    output.default?.steps.forEach((step) => {
      delete step.unikey
    })
    expect(output.default).toEqual(res)
    const output2 = await getConfig('a', 'cwf.config3.js', __dirname)
    // remove unikey
    output2.default?.steps.forEach((step) => {
      delete step.unikey
    })
    const output2Res = {
      logLevel: 'info',
      isSkipError: false,
      isThrowErrorBreak: false,
      steps: [{
        command: 'git tag v#{tag} -m #{message}',
        tags: {
          tag: 'tag',
          message: 'message',
        },
      }, {
        command: 'ls',
      }],
    }
    expect(output2.default).toEqual(output2Res)
  })
})
