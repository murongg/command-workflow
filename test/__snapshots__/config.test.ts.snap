// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`describe > test defineConfig 1`] = `
{
  "logLevel": "info",
  "steps": [
    {
      "command": "git tag v#{tag} -m #{message}",
      "tags": {
        "message": "message",
        "tag": "tag",
      },
    },
  ],
}
`;
